`skaffold fix` only seems to work with the latest schema version

```
$ skaffold version
v1.11.0
$ cat skaffold.yaml
apiVersion: skaffold/v1beta15
kind: Config
build:
  tagPolicy:
    sha256: {}
  # defines where to find the code at build time and where to push the resulting image
  artifacts:
    - image: java-hello-world
      # To learn more about how Jib builds Java containers visit
      # https://github.com/GoogleContainerTools/jib
      jib: {}

# defines the Kubernetes manifests to deploy on each run
deploy:
  kubectl:
    manifests:
    - ./kubernetes-manifests/**.yaml
profiles:
# use the cloudbuild profile to build images using Google Cloud Build
- name: cloudbuild
  build:
    googleCloudBuild: {}
# use the dockerfile profile to build images using Docker instead of Jib
- name: dockerfile
  build:
    artifacts:
      - image: java-hello-world
$ skaffold fix --version='skaffold/v2beta4'
panic: interface conversion: util.VersionedConfig is *v2beta4.SkaffoldConfig, not *latest.SkaffoldConfig

goroutine 1 [running]:
github.com/GoogleContainerTools/skaffold/cmd/skaffold/app/cmd.fix(0x5ce6300, 0xc00020e008, 0x59816c6, 0xd, 0x7ffeefbffa9f, 0x10, 0xc00020e000, 0x5ce6300, 0xc00020e008)
	/private/tmp/skaffold-20200612-72105-1hzc90j/src/github.com/GoogleContainerTools/skaffold/cmd/skaffold/app/cmd/fix.go:70 +0x52f
github.com/GoogleContainerTools/skaffold/cmd/skaffold/app/cmd.doFix(0x5d2dd80, 0xc00005c040, 0x5ce6300, 0xc00020e008, 0xc00020e008, 0x5978ee4)
	/private/tmp/skaffold-20200612-72105-1hzc90j/src/github.com/GoogleContainerTools/skaffold/cmd/skaffold/app/cmd/fix.go:51 +0x74
github.com/GoogleContainerTools/skaffold/cmd/skaffold/app/cmd.(*builder).NoArgs.func1(0xc0000d18c0, 0xc0006b5860, 0x0, 0x1, 0x0, 0x0)
	/private/tmp/skaffold-20200612-72105-1hzc90j/src/github.com/GoogleContainerTools/skaffold/cmd/skaffold/app/cmd/commands.go:99 +0xbb
github.com/spf13/cobra.(*Command).execute(0xc0000d18c0, 0xc0006b5850, 0x1, 0x1, 0xc0000d18c0, 0xc0006b5850)
	/private/tmp/skaffold-20200612-72105-1hzc90j/src/github.com/GoogleContainerTools/skaffold/vendor/github.com/spf13/cobra/command.go:842 +0x453
github.com/spf13/cobra.(*Command).ExecuteC(0xc0000d0000, 0xc00020e008, 0x5ce6300, 0xc00020e010)
	/private/tmp/skaffold-20200612-72105-1hzc90j/src/github.com/GoogleContainerTools/skaffold/vendor/github.com/spf13/cobra/command.go:950 +0x349
github.com/spf13/cobra.(*Command).Execute(...)
	/private/tmp/skaffold-20200612-72105-1hzc90j/src/github.com/GoogleContainerTools/skaffold/vendor/github.com/spf13/cobra/command.go:887
github.com/spf13/cobra.(*Command).ExecuteContext(...)
	/private/tmp/skaffold-20200612-72105-1hzc90j/src/github.com/GoogleContainerTools/skaffold/vendor/github.com/spf13/cobra/command.go:880
github.com/GoogleContainerTools/skaffold/cmd/skaffold/app.Run(0x5ce6300, 0xc00020e008, 0x5ce6300, 0xc00020e010, 0x0, 0x0)
	/private/tmp/skaffold-20200612-72105-1hzc90j/src/github.com/GoogleContainerTools/skaffold/cmd/skaffold/app/skaffold.go:33 +0xde
main.main()
	/private/tmp/skaffold-20200612-72105-1hzc90j/src/github.com/GoogleContainerTools/skaffold/cmd/skaffold/skaffold.go:31 +0x4e
$ skaffold fix --version='skaffold/v2beta5'
apiVersion: skaffold/v2beta5
kind: Config
build:
  artifacts:
  - image: java-hello-world
    jib: {}
  tagPolicy:
    sha256: {}
deploy:
  kubectl:
    manifests:
    - ./kubernetes-manifests/**.yaml
profiles:
- name: cloudbuild
  build:
    googleCloudBuild: {}
- name: dockerfile
  build:
    artifacts:
    - image: java-hello-world
```

I think these lines are assuming we're always upgrading to the latest version, instead of version given by `--version`: https://github.com/GoogleContainerTools/skaffold/blob/master/cmd/skaffold/app/cmd/fix.go#L70-L72

--- a/cmd/skaffold/app/cmd/fix.go
+++ b/cmd/skaffold/app/cmd/fix.go
@@ -57,8 +57,8 @@ func fix(out io.Writer, configFile string, toVersion string, overwrite bool) err
 		return err
 	}
 
-	if cfg.GetVersion() == latest.Version {
-		color.Default.Fprintln(out, "config is already latest version")
+	if cfg.GetVersion() == toVersion {
+		color.Default.Fprintln(out, "config is already version", toVersion)
 		return nil
 	}
 
@@ -67,8 +67,12 @@ func fix(out io.Writer, configFile string, toVersion string, overwrite bool) err
 		return err
 	}
 
-	if err := validation.Process(cfg.(*latest.SkaffoldConfig)); err != nil {
-		return fmt.Errorf("validating upgraded config: %w", err)
+	// TODO(dgageot): We should be able run validations on any schema version
+	// but that's not the case. They can only run on the latest version for now.
+	if toVersion == latest.Version {
+		if err := validation.Process(cfg.(*latest.SkaffoldConfig)); err != nil {
+			return fmt.Errorf("validating upgraded config: %w", err)
+		}
 	}
 
 	newCfg, err := yaml.Marshal(cfg)

--- a/cmd/skaffold/app/cmd/fix_test.go
+++ b/cmd/skaffold/app/cmd/fix_test.go
@@ -23,18 +23,21 @@ import (
 	"testing"
 
 	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/schema/latest"
+	v1 "github.com/GoogleContainerTools/skaffold/pkg/skaffold/schema/v1"
 	"github.com/GoogleContainerTools/skaffold/testutil"
 )
 
 func TestFix(t *testing.T) {
 	tests := []struct {
-		description string
-		inputYaml   string
-		output      string
-		shouldErr   bool
+		description   string
+		inputYaml     string
+		targetVersion string
+		output        string
+		shouldErr     bool
 	}{
 		{
-			description: "v1alpha4 to latest",
+			description:   "v1alpha4 to latest",
+			targetVersion: latest.Version,
 			inputYaml: `apiVersion: skaffold/v1alpha4
 kind: Config
 build:
@@ -69,7 +72,8 @@ deploy:
 `, latest.Version),
 		},
 		{
-			description: "v1alpha1 to latest",
+			description:   "v1alpha1 to latest",
+			targetVersion: latest.Version,
 			inputYaml: `apiVersion: skaffold/v1alpha1
 kind: Config
 build:
@@ -96,16 +100,57 @@ deploy:
 `, latest.Version),
 		},
 		{
-			description: "already latest version",
+			description:   "v1alpha1 to v1",
+			targetVersion: v1.Version,
+			inputYaml: `apiVersion: skaffold/v1alpha1
+kind: Config
+build:
+  artifacts:
+  - imageName: docker/image
+    dockerfilePath: dockerfile.test
+deploy:
+  kubectl:
+    manifests:
+    - paths:
+      - k8s/deployment.yaml
+`,
+			output: fmt.Sprintf(`apiVersion: %s
+kind: Config
+build:
+  artifacts:
+  - image: docker/image
+    docker:
+      dockerfile: dockerfile.test
+deploy:
+  kubectl:
+    manifests:
+    - k8s/deployment.yaml
+`, v1.Version),
+		},
+		{
+			description:   "already target version",
+			targetVersion: latest.Version,
 			inputYaml: fmt.Sprintf(`apiVersion: %s
 kind: Config
 `, latest.Version),
-			output: "config is already latest version\n",
+			output: "config is already version " + latest.Version + "\n",
 		},
 		{
 			description: "invalid input",
 			inputYaml:   "invalid",
 			shouldErr:   true,
+		},
+		{
+			description:   "validation fails",
+			targetVersion: latest.Version,
+			inputYaml: `apiVersion: skaffold/v1alpha1
+kind: Config
+build:
+  artifacts:
+  - imageName: 
+    dockerfilePath: dockerfile.test
+`,
+			shouldErr: true,
 		},
 	}
 	for _, test := range tests {
@@ -113,7 +158,7 @@ kind: Config
 			cfgFile := t.TempFile("config", []byte(test.inputYaml))
 
 			var b bytes.Buffer
-			err := fix(&b, cfgFile, latest.Version, false)
+			err := fix(&b, cfgFile, test.targetVersion, false)
 
 			t.CheckErrorAndDeepEqual(test.shouldErr, err, test.output, b.String())
 		})

